### PR TITLE
Add readers to consume tracepoints with callstacks

### PR DESCRIPTION
**Pull request by Mahmood-Darwish**
This PR adds the readers that will be used to read the new type of events for tracepoint call stack collection.

Bug: http://b/242825690

### CODE REVIEW

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 1]
```diff
@@ -445,6 +445,150 @@ SchedWakeupPerfEvent ConsumeSchedWakeupPerfEvent(PerfEventRingBuffer* ring_buffe
...
+  // tracepoint_with_callchain_event_open
+  const perf_event_attr flags{.sample_type = PERF_SAMPLE_CALLCHAIN | PERF_SAMPLE_RAW |
+                                             SAMPLE_TYPE_TID_TIME_STREAMID_CPU |
+                                             PERF_SAMPLE_REGS_USER | PERF_SAMPLE_STACK_USER,
+                              .sample_regs_user = SAMPLE_REGS_USER_ALL};
+
+  PerfRecordSample res = ConsumeRecordSample(ring_buffer, header, flags);
+
+  sched_wakeup_tracepoint_fixed sched_wakeup;
+  std::memcpy(&sched_wakeup, res.raw_data.get(), sizeof(sched_wakeup_tracepoint_fixed));
```
Comment (florian-kuebler) [inline src/LinuxTracing/PerfEventReaders.cpp:461]: It is not a lot of data, but I am wondering if we should avoid the memcpy and just do:
```
const auto* sched_wakeup = absl::bitcast<const sched_wakeup_tracepoint_fixed*>(res.raw_data.get());
```

Comment (florian-kuebler) [reply src/LinuxTracing/PerfEventReaders.cpp:461]: Same below and in the existing cases. WDYT?

File: src/LinuxTracing/PerfEventReaders.h [hunk 2]
```diff
@@ -49,6 +49,18 @@ GenericTracepointPerfEvent ConsumeGenericTracepointPerfEvent(PerfEventRingBuffer
 SchedWakeupPerfEvent ConsumeSchedWakeupPerfEvent(PerfEventRingBuffer* ring_buffer,
                                                  const perf_event_header& header);
 
+SchedWakeupWithCallchainPerfEvent ConsumeSchedWakeupWithCallchainPerfEvent(
```
Comment (florian-kuebler) [inline src/LinuxTracing/PerfEventReaders.h:52]: can we take the chance to add `[[nodiscard]]` to all the read methods?

**Review by florian-kuebler — APPROVED** [review]
Only minor comments. Besides that, lgtm. Thanks a lot!

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 1 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:461]: I'm not really sure which is better. Can we even do it for all cases? One of the structs (I think the `sched_wakeup_tracepoint_fixed` has holds less data than `raw_data`) we just copy a part of `raw_data` to fill the struct and leave the rest. Would that still work with the bit case?

I have no preferences, @beckerhe  what do you think?

Comment (beckerhe) [reply src/LinuxTracing/PerfEventReaders.cpp:461]: Using bit_cast on a pointer is basically reinterpret_cast. From the bit_cast docs:

```
Using bit_cast on a pointer and then dereferencing it is no better than using
// reinterpret_cast. You should only use bit_cast on the value itself.
```

Using memcpy is the only valid way to do this.

File: src/LinuxTracing/PerfEventReaders.h (hunk 2 above)
Comment (beckerhe) [reply src/LinuxTracing/PerfEventReaders.h:52]: +1

**Review by beckerhe — APPROVED** [review]
Awesome. Looks all good to me!

File: src/LinuxTracing/PerfEventReaders.h (hunk 2 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.h:52]: Done.

**Pull request merged**